File: ios/Readium/ReadiumViewManager.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(ReadiumViewManager, RCTViewManager)

// Shared
RCT_EXPORT_VIEW_PROPERTY(file, NSDictionary *)
RCT_EXPORT_VIEW_PROPERTY(location, NSDictionary *)
RCT_EXPORT_VIEW_PROPERTY(onTableOfContents, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onReady, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDownloaded, RCTDirectEventBlock)

// Reader
RCT_EXPORT_VIEW_PROPERTY(settings, NSDictionary *)
RCT_EXPORT_VIEW_PROPERTY(onLocationChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onFullModeChange, RCTDirectEventBlock)
RCT_EXTERN_METHOD(goToPosition:(nonnull NSNumber*) reactTag
                  position:(CGFloat*) position)
RCT_EXTERN_METHOD(goToLocation:(nonnull NSNumber*) reactTag
                  location:(NSDictionary*) position)

// Audio
RCT_EXTERN_METHOD(playAudio:(nonnull NSNumber*) reactTag)
RCT_EXTERN_METHOD(pauseAudio:(nonnull NSNumber*) reactTag)
RCT_EXTERN_METHOD(seekAudio:(nonnull NSNumber*) reactTag value:(CGFloat*) value)
RCT_EXTERN_METHOD(setAudioRate:(nonnull NSNumber*) reactTag rate:(CGFloat*) rate)
RCT_EXTERN_METHOD(goToNextAudio:(nonnull NSNumber*) reactTag)
RCT_EXTERN_METHOD(goToPreviousAudio:(nonnull NSNumber*) reactTag)
RCT_EXTERN_METHOD(goToAudioPosition:(nonnull NSNumber*) reactTag
                  position:(CGFloat*) position)
RCT_EXPORT_VIEW_PROPERTY(onAudioPlaybackChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDownloading, RCTDirectEventBlock)
@end

File: ios/Readium/ReadiumViewManager.swift
import Foundation

@objc(ReadiumViewManager)
class ReadiumViewManager: RCTViewManager {
  override func view() -> (ReadiumView) {
    return ReadiumView()
  }

  override class func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc func goToPosition(_ reactTag: NSNumber, position: Double) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.goToPosition(position: position)
    }
  }
  
  @objc func goToLocation(_ reactTag: NSNumber, location: NSDictionary) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.goToLocation(location: location)
    }
  }
  
  @objc func playAudio(_ reactTag: NSNumber) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.playAudio()
    }
  }
  
  @objc func pauseAudio(_ reactTag: NSNumber) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.pauseAudio()
    }
  }
  
  @objc func seekAudio(_ reactTag: NSNumber, value: Double) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.seekAudio(value: value)
    }
  }
  
  @objc func goToNextAudio(_ reactTag: NSNumber) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.goToNextAudio()
    }
  }
  
  @objc func goToPreviousAudio(_ reactTag: NSNumber) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.goToPreviousAudio()
    }
  }
  
  @objc func goToAudioPosition(_ reactTag: NSNumber, position: Double) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.goToAudioPosition(position: position)
    }
  }
  
  @objc func setAudioRate(_ reactTag: NSNumber, rate: Double) {
    self.bridge!.uiManager.addUIBlock { (_: RCTUIManager?, viewRegistry: [NSNumber: UIView]?) in
        let view: ReadiumView = (viewRegistry![reactTag] as? ReadiumView)!
        view.setAudioRate(rate: rate)
    }
  }
}
